InstanceManager.js

```diff
@@ -647,97 +730,88 @@ class InstanceManager {
+        throw new Error(instance.name + ' did not stop gracefully after ' + (waitInterval * attempts) + 'ms');
+      }
+
+      // wait a while and try again.
+      await sleep(shutdownWaitInterval);
     }
+  }
 
-    let checkDown = function() {
-      return new Promise((resolve, reject) => {
-        if (process.env.LOG_IMMEDIATE && process.env.LOG_IMMEDIATE == "1") {
-          console.log((new Date()).toISOString()
-            + " checkDown called for " + instance.name);
-        }
-        let attempts = 0;
-        let killAttempts = 3600;  // 180s, after this time we kill the instance
-        let maxAttempts = 4000;   // 200s, note that the cluster internally
-                                  // has a 120s timeout
-        let waitInterval = 50;
-        (function innerCheckDown() {
-          if (instance.status == 'EXITED') {
-            if (process.env.LOG_IMMEDIATE && process.env.LOG_IMMEDIATE == "1") {
-              console.log((new Date()).toISOString()
-                + " innerCheckDown resolve for " + instance.name);
-            }
-            resolve(instance);
-          } else if (++attempts === killAttempts) {
-            if (process.env.LOG_IMMEDIATE && process.env.LOG_IMMEDIATE == "1") {
-              console.log((new Date()).toISOString()
-                + " innerCheckDown: killed" + instance.name);
-            }
-            instance.process.kill('SIGKILL');
-            instance.status = 'KILLED';
-            setTimeout(innerCheckDown, waitInterval);
-          } else if (attempts > maxAttempts) {
-            if (process.env.LOG_IMMEDIATE && process.env.LOG_IMMEDIATE == "1") {
-              console.log((new Date()).toISOString()
-                + " innerCheckDown reject for " + instance.name);
-            }
-            reject(new Error(instance.name + ' did not stop gracefully after ' + (waitInterval * attempts) + 'ms'));
-          } else {
-            setTimeout(innerCheckDown, waitInterval);
-          }
-        })();
-      });
+  async shutdown(instance) {
+    if (instance.status == 'EXITED') {
+      return instance;
     }
+    debugLog(`Shutdown ${instance.name}`);
 
-    return rp.delete({
-      url: this.getEndpointUrl(instance) + '/_admin/shutdown',
-    })
-    .catch(err => {
+    try {
+      await rp.delete({ url: this.getEndpointUrl(instance) + '/_admin/shutdown' })
+    } catch (err) {
+      let expected = false;
       if (err && (err.statusCode === 503 || err.error)) {
-        if (err.error.code == 'ECONNREFUSED') {
-          console.warn('hmmm...server ' + instance.name + ' did not respond (' + err.code + '). Assuming it is dead. Status is: ' + instance.status);
-          return checkDown();
-        } else if (err.error.code == 'ECONNRESET') {
-          return checkDown();
+        // Some "errors" are expected.
+        let errObj = (err.error instanceof String || typeof err.error === "string") ?
```

Comment:
`instanceof String` is almost never what you want. Just use `typeof` if you want to check for a string.